# **[윤슬🌊] 퇴근시간 버스승차인원 예측 경진대회**
# **Stage 1: 데이터 불러오기**

![image](http://www.jejunews.com/news/photo/202003/2159886_183759_925.jpg)


안녕하세요. 데이크루 4기 '윤슬' 팀입니다!
</br>저희는 퇴근시간 버스승차인원 예측 프로젝트의 데이콘 문제 기반 학습(Problem Based Learning, PBL)을 만들고 있습니다.

첫번째 포스팅의 주제는 <font style = "color:#6495ED; font-weight:bold; font-size:25px">'데이터 불러오기'</font>입니다.

*본 포스팅은 데이콘 서포터즈 '데이크루' 4기 활동의 일환입니다.*
<p style='text-align: right;'> 23-01-19|ver.001</p>

# **1. Intro**

매일 아침 집을 나갈 때마다 <font style = "color:#6495ED; font-weight:bold; font-size:25px">'퇴근을 바라는 것'</font>이 모든 직장인들의 마음이죠?!
힘든 몸을 이끌고 집에 가는 버스에 탈 때마다 항상 꽉꽉 차있는 만원 버스!
</br>이런 만원버스에서는 어떤 일이 벌어지고 있을까요?
초보부터 고수까지 모두가 참여하는 퇴근시간 버스승차인원 예측 프로젝트를 소개합니다!🚌
 
퇴근시간 버스승차인원 예측에 참여하신 데이커 여러분, 환영합니다.
 
직장인의 가장 큰 소원은 퇴근이죠.
퇴근시간 버스승차인원이 적기를 바라는건 당연한 마음!
 
이번 교육을 통해 퇴근시간 버스승차인원 데이터를 분석하여 직접 타보지 않고도 승차 인원을 예측해 내는 멋진 데이커가 되어 봅시다.✨

## **1.1 프로젝트 배경**
![image](https://ifh.cc/g/WXxpht.jpg)
[그림1] “버스” 이모티콘에 대한 추천만으로도 버스 승차 인원이 만만치 않다는 것을 확인할 수 있습니다.
</br><font style = "color:gray;"><출처>카카오톡 이모티콘 플러스</font>

매일 빠르게 집에 가고 싶지만, 항상 퇴근하고 집에만 가는 것은 아닙니다.
</br>친구와 약속을 잡기도 하고 반차를 써서 일찍 퇴근하기도 하죠.
</br>이처럼 장소나 시간에 관한 고민 역시 퇴근시간 버스승차 인원을 결정하는 중요한 요인 중 하나입니다.
 
버스가 오는 걸 보지 않고도 오늘 몇명이나 버스에 탑승할지 예측할 수 있다면, 이런 고민을 덜 수 있겠죠!
 
그럼 퇴근시간 버스승차인원 예측 프로젝트, 지금 바로 시작해 볼까요?

# **2. 데이터 살펴보기**
## **2.1 데이터 불러오기**
데이터 분석을 하기 위한 첫 번째 단계는 무엇일까요?

바로 데이터를 살펴보는 일입니다. 주어진 데이터를 살펴보며 데이터의 모양과 대략적인 내용을 파악하고, 어떤 분석이 가능한지 확인하여 방향을 잡는 과정이 필요합니다. 

편리하게 데이터를 확인하기 위해선 파이썬 패키지를 잘 활용해야 하는데, 이번엔 <font style = "color:#6495ED; font-weight:bold; font-size:25px">판다스(Pandas)</font>를 사용해보도록 하겠습니다. 
</br>pandas는 "관계형" 또는 "라벨이 있는" 데이터를 쉽고 직관적으로 사용할 수 있도록 설계된 빠르고 유연하며 표현력이 뛰어난 데이터 구조를 제공합니다. 
</br>즉, pandas는 다양한 종류의 데이터에 매우 적합합니다.

pandas 패키지를 사용하기 위해 우선 임포트를 해야 합니다. “import” 패키지 “as” 약어로 이루어져 있습니다. 
</br>import 명령어는 설치된 pandas 패키지를 불러와서 실행할 수 있게 대기 상태로 만들어 주는 명령어입니다.

In [1]:
# 데이터를 불러오고 살펴보기 위한 pandas 라이브러리
import pandas as pd

이번 프로젝트의 데이터는 총 4가지이며 모두 csv파일입니다.

1. train.csv : 데이터 분석 모델을 학습(train)하기 위해 사용하는 데이터 </br> 2019년 9월 제주도의 각 날짜, 출근시간(6시 ~ 12시)의 버스 정류장별 승차인원, 퇴근시간(18시 ~ 20시)의 버스 정류장별 승차인원
2. test.csv : 데이터 분석 모델을 통해 정답을 예측하기 위해 사용하는 데이터</br>2019년 10월의 각 날짜, 출근시간(6시 ~ 12시)의 버스 정류장별 승하차 인원
3. sample_submission.csv : 예측한 정답 값을 기록하여 제출하기 위한 샘플 정답 파일
4. bus_bts : 데이터 분석 모델을 위한 추가적인 데이터 탑승 시간대가 6시부터 12시 사이인 버스카드별 승하차 정보

분석에 사용한 데이터 파일을 불러오기 위해, pandas 패키지가 가지고 있는 read_csv() 함수를 사용하겠습니다.

분석에 사용할 데이터를 한번 불러와보겠습니다.

`read_csv()` 함수는 (‘파일 경로/파일명.확장자명’) 의 형식을 받아 데이터를 불러옵니다.

코드가 실행되고 있는 파일(.ipynb)과 불러오고 싶은 데이터 파일(.csv)이 동일한 위치에 존재하는 경우, </br>read_csv() 함수는 파일 경로를 생략하고 (‘파일명.확장자명’) 의 형식을 받아 데이터를 불러올 수 있습니다.

이때 index_col 옵션을 이용하면 인덱스 컬럼을 제거할 수 있습니다. 즉, index_col 값을 0으로 지정하면 Unnamed: 0 컬럼이 없어집니다. 

In [2]:
# train이라는 객체에 'train.csv' 파일을 할당해 줍니다. 
train = pd.read_csv('train.csv' ,index_col=0)

# submission이라는 객체에 'sample_submission.csv' 파일을 할당해 줍니다. 
submission = pd.read_csv('submission_sample.csv',index_col=0)

### **🔑참여 실습**


### *Inst.*  
입력되어 있는 코드셀을 확인하고, test.csv파일을 직접 불러와봅시다.






### *Hint.*  
pd.read_csv() 




### *Solution.*

In [3]:
test = pd.read_csv('test.csv', index_col=0)

## **2.2 데이터 확인**

데이터가 잘 불러와졌는지, 어떻게 생겼는지 살펴봅시다.

데이터 과학 실무 과정에서는 데이터를 살펴볼 때 head(), tail(), info() 메소드를 활용하는 경우가 많습니다. 

head() : 상위 5행을 선택
</br> tail() : 하위 5행을 선택
</br> info(): 데이터 결측치 및 변수들의 타입 확인

데이터를 정렬하여 이상치나 결측치와 같은 데이터가 있는지 보기 위함입니다. 
이 두 가지 메소드를 이용해서 데이터가 어떻게 생겼는지 확인해봅시다.

In [4]:
# train 데이터 상위 5개 행 출력
train.head()

date  bus_route_id in_out  station_code  station_name  latitude  \
id                                                                          
0   2019-09-01       4270000     시외           344         제주썬호텔  33.48990   
1   2019-09-01       4270000     시외           357          한라병원  33.48944   
2   2019-09-01       4270000     시외           432          정존마을  33.48181   
3   2019-09-01       4270000     시내          1579  제주국제공항(600번)  33.50577   
4   2019-09-01       4270000     시내          1646      중문관광단지입구  33.25579   

    longitude  6~7_ride  7~8_ride  8~9_ride  9~10_ride  10~11_ride  \
id                                                                   
0   126.49373       0.0       1.0       2.0        5.0         2.0   
1   126.48508       1.0       4.0       4.0        2.0         5.0   
2   126.47352       1.0       1.0       0.0        2.0         0.0   
3   126.49252       0.0      17.0       6.0       26.0        14.0   
4   126.41260       0.0       0.0       0.0        0.0         0.0   

    11~12_ride  6~7_takeoff  7~8_takeoff  8~9_takeoff  9~10_takeoff  \
id                                                                    
0          6.0          0.0          0.0          0.0           0.0   
1          6.0          0.0          0.0          0.0           0.0   
2          0.0          0.0          0.0          0.0           0.0   
3         16.0          0.0          0.0          0.0           0.0   
4          0.0          0.0          0.0          0.0           1.0   

    10~11_takeoff  11~12_takeoff  18~20_ride  
id                                            
0             0.0            0.0         0.0  
1             0.0            0.0         5.0  
2             0.0            0.0         2.0  
3             0.0            0.0        53.0  
4             0.0            0.0         0.0

In [5]:
# train 데이터 결측치 및 변수들의 타입 확인
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 415423 entries, 0 to 415422
Data columns (total 20 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   date           415423 non-null  object 
 1   bus_route_id   415423 non-null  int64  
 2   in_out         415423 non-null  object 
 3   station_code   415423 non-null  int64  
 4   station_name   415423 non-null  object 
 5   latitude       415423 non-null  float64
 6   longitude      415423 non-null  float64
 7   6~7_ride       415423 non-null  float64
 8   7~8_ride       415423 non-null  float64
 9   8~9_ride       415423 non-null  float64
 10  9~10_ride      415423 non-null  float64
 11  10~11_ride     415423 non-null  float64
 12  11~12_ride     415423 non-null  float64
 13  6~7_takeoff    415423 non-null  float64
 14  7~8_takeoff    415423 non-null  float64
 15  8~9_takeoff    415423 non-null  float64
 16  9~10_takeoff   415423 non-null  float64
 17  10~11_takeoff  415423 non-nul

train.info()의 결과를 보면 Dtype부분에 object와 int64 그리고 float64가 존재합니다. 
</br> int64는 정수, float64는 실수를 의미하고 object는 문자열 객체를 의미합니다.

이번에는 <font style = "color:#6495ED; font-weight:bold; font-size:25px">넘파이(Numpy)</font>라는 패키지를 불러올겁니다. 
</br> 앞서 pandas를 불러왔던 것과 동일한 방식을 사용할 겁니다. 


In [6]:
import numpy as np

이 Numpy는 다차원 배열을 쉽게 처리하고 효율적으로 사용할 수 있도록 지원하는 파이썬의 패키지입니다. 
</br> NumPy는 데이터 구조 외에도 수치 계산을 위해 효율적으로 구현된 기능을 제공합니다. 데이터 분석을 할때 자주 사용되는 패키지 중 하나입니다.

NumPy 배열에서 배열의 형태를 알아보는 shape()란 함수가 있습니다.
</br>즉, shape은 해당 데이터의 차원이 몇 차원인지 표시해줍니다.

In [7]:
np.shape(train)

(415423, 20)

train 데이터는 415423x20의 배열 형태를 가지고 있다는 걸 알 수 있습니다. 

train 데이터를 살펴보니 총 20개의 열이 있네요. 하나하나 살펴보겠습니다. 

1. id : 해당 데이터에서의 고유한 ID
2. date : 버스 운행 날짜
3. bus_route_id: 버스 노선 ID
4. in_out: 시내버스, 시외버스 구분
5. station_code: 해당 승하차 정류소의 ID
6. station_name: 해당 승하차 정류소의 이름
7. latitude: 해당 버스 정류장의 위도( 같은 정류장 이름이어도 버스의 진행 방향에 따라 다를 수 있음)
8. longitude: 해당 버스 정류장의 경도( 같은 정류장 이름이어도 버스의 진행 방향에 따라 다를 수 있음)
9. 6~7_ride: 6시부터 6시 59분 59초까지 승차한 인원 수
10. 7~8_ride: 7시부터 7시 59분 59초까지 승차한 인원 수
11. 8~9_ride: 8시부터 8시 59분 59초까지 승차한 인원 수
12. 9~10_ride: 9시부터 9시 59분 59초까지 승차한 인원 수
13. 10~11_ride: 10시부터 10시 59분 59초까지 승차한 인원 수
14. 11~12_ride: 11시부터 11시 59분 59초까지 승차한 인원 수
15. 6~7_takeoff: 6시부터 6시 59분 59초까지 하차한 인원 수
16. 7~8_takeoff: 7시부터 7시 59분 59초까지 하차한 인원 수
17. 8~9_takeoff: 8시부터 8시 59분 59초까지 하차한 인원 수
18. 9~10_takeoff: 9시부터 9시 59분 59초까지 하차한 인원 수
19. 10~11_takeoff: 10시부터 10시 59분 59초까지 하차한 인원 수
20. 11~12_takeoff: 11시부터 11시 59분 59초까지 하차한 인원 수
21. 18~20_ride: 18시부터 19시 59분 59초까지 승차한 인원의 수 ( train data에만 존재)

</br> 총 21개의 변수와  415423개의 행으로 이루어져 있습니다. 
</br> 또한 Non-Null Count를 살펴보면 결측치는 존재하지 않는다는 것을 알 수 있습니다. 


### 🔑**참여 실습**


### *Inst.*  
다음으로 test 데이터도 살펴봐봅시다. test데이터의 하위 5개 행을 출력해봅시다.






### *Hint.*  
tail()






### *Solution.* 

test.tail()

### 🔑**참여 실습**


### *Inst.*  
test 데이터 결측치 및 변수들의 타입을 확인해봅시다. 






### *Hint.*  
info()






### *Solution.* 

In [9]:
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 228170 entries, 415423 to 643592
Data columns (total 19 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   date           228170 non-null  object 
 1   bus_route_id   228170 non-null  int64  
 2   in_out         228170 non-null  object 
 3   station_code   228170 non-null  int64  
 4   station_name   228170 non-null  object 
 5   latitude       228170 non-null  float64
 6   longitude      228170 non-null  float64
 7   6~7_ride       228170 non-null  float64
 8   7~8_ride       228170 non-null  float64
 9   8~9_ride       228170 non-null  float64
 10  9~10_ride      228170 non-null  float64
 11  10~11_ride     228170 non-null  float64
 12  11~12_ride     228170 non-null  float64
 13  6~7_takeoff    228170 non-null  float64
 14  7~8_takeoff    228170 non-null  float64
 15  8~9_takeoff    228170 non-null  float64
 16  9~10_takeoff   228170 non-null  float64
 17  10~11_takeoff  228170 no

test 데이터에는 18-20_ride 변수가 빠지고 총 19개의 열이 존재합니다. 
</br> 그러므로 이번 분석은 18-20_ride, 즉 18시부터 20시까지 승차한 인원의 수를 예측하는 문제라고 파악할 수 있습니다. 

### 🔑**참여 실습**


### *Inst.*  
test 데이터의 (행, 열) 크기를 출력해봅시다.






### *Hint.*  
shape()






### *Solution.* 

In [10]:
np.shape(test)

(228170, 19)

마지막으로 submission 데이터도 한번 살펴봅시다.

# **3. 제출파일 생성**

앞서 sample_submission.csv 파일을 샘플 정답 파일이라고 소개했었습니다.
</br> 앞으로는 이 sample_submission.csv 파일을 이용해서 제출 파일을 만들 예정입니다. 
</br> 그러면 제출 파일을 생성하는 방법을 간단하게 알아봅시다. 
</br> </br> 18-20_ride 값을 일단 train 데이터의 18~20_ride의 평균으로 채워봅시다. 

In [11]:
submission['18~20_ride'] = train['18~20_ride'].mean()
submission.head()

18~20_ride
id                
415423    1.242095
415424    1.242095
415425    1.242095
415426    1.242095
415427    1.242095

이제 평균으로 채운 submission파일을 submission.csv에 저장하고 제출해봅시다.
</br>기본적으로 to_csv 메소드는 데이터 프레임 객체의 내용을 csv 파일 형태로 만들어서 저장해 주는 기능입니다. 
</br>이렇게 해야 파이썬 안에서 작업한 내용이 파이썬 밖으로 빠져나가 여러분이 만질 수 있는 형태가 되는 겁니다. 

to_csv 메서도 안에 아무런 옵션도 지정해주지 않는다면 다음과 같이 인덱스도 데이터에 포함되어 저장되기 떄문에 </br>index는 포함하지않고 저장하는 옵션을 지정해주어야 합니다.
 
index를 포함시키지 않고 데이터를 저장하기 위해서는 to_csv(inedx=False)로 지정해서 데이터를 저장해주면 됩니다.

### 🔑**참여 실습**


### *Inst.*  
submission파일을 저장해봅시다.






### *Hint.*  
to_csv()






### *Solution.* 

In [12]:
submission.to_csv('submission.csv', index=False)